src/storage/compression/zstd.cpp

```diff
@@ -0,0 +1,1054 @@
+		context = duckdb_zstd::ZSTD_createCCtx();
+	}
+	~ZSTDAnalyzeState() override {
+		duckdb_zstd::ZSTD_freeCCtx(context);
+	}
+
+public:
+	inline void AppendString(const string_t &str) {
+		auto string_size = str.GetSize();
+		total_size += string_size;
+	}
+
+public:
+	DBConfig &config;
+
+	duckdb_zstd::ZSTD_CCtx *context;
+	//! The combined string lengths for all values in the segment
+	idx_t total_size = 0;
+	//! The total amount of values in the segment
+	idx_t count = 0;
+
+	//! The amount of vectors per filled segment
+	idx_t vectors_per_segment = 0;
+	//! The total amount of segments we will create
+	idx_t segment_count = 0;
+	//! Current vector in the segment
+	idx_t vectors_in_segment = 0;
+	//! Current amount of values in the vector
+	idx_t values_in_vector = 0;
+};
+
+unique_ptr<AnalyzeState> ZSTDStorage::StringInitAnalyze(ColumnData &col_data, PhysicalType type) {
+	CompressionInfo info(col_data.GetBlockManager().GetBlockSize());
+	auto &data_table_info = col_data.info;
+	auto &attached_db = data_table_info.GetDB();
+	auto &config = DBConfig::Get(attached_db);
+
+	return make_uniq<ZSTDAnalyzeState>(info, config);
+}
+
+bool ZSTDStorage::StringAnalyze(AnalyzeState &state_p, Vector &input, idx_t count) {
+	auto &state = state_p.Cast<ZSTDAnalyzeState>();
+	UnifiedVectorFormat vdata;
+	input.ToUnifiedFormat(count, vdata);
+
+	auto data = UnifiedVectorFormat::GetData<string_t>(vdata);
+	for (idx_t i = 0; i < count; i++) {
+		auto idx = vdata.sel->get_index(i);
+		if (!vdata.validity.RowIsValid(idx)) {
+			continue;
+		}
+		auto &str = data[idx];
+		auto string_size = str.GetSize();
+		state.total_size += string_size;
+	}
+	state.values_in_vector += count;
+	while (state.values_in_vector >= ZSTD_VECTOR_SIZE) {
+		if (GetVectorMetadataSize(state.vectors_in_segment + 1) > GetWritableSpace(state.info)) {
+			state.vectors_per_segment = state.vectors_in_segment;
+			state.segment_count++;
+			state.vectors_in_segment = 0;
+		} else {
+			state.vectors_in_segment++;
+		}
+		state.values_in_vector -= ZSTD_VECTOR_SIZE;
+	}
+	state.count += count;
+	return true;
+}
+
+// Compression score to determine which compression to use
+idx_t ZSTDStorage::StringFinalAnalyze(AnalyzeState &state_p) {
+	auto &state = state_p.Cast<ZSTDAnalyzeState>();
+
+	if (!state.count) {
+		return NumericLimits<idx_t>::Maximum();
+	}
+
+	if (state.values_in_vector) {
+		D_ASSERT(state.values_in_vector < ZSTD_VECTOR_SIZE);
+		state.segment_count++;
+	}
+
+	double penalty;
+	idx_t average_length = state.total_size / state.count;
+	auto overflow_string_threshold = StringUncompressed::GetStringBlockLimit(state.info.GetBlockSize());
+	if (average_length < overflow_string_threshold) {
+		average_length = MaxValue<idx_t>(average_length, 200);
+		penalty =
+		    5.0 +
+		    (((double)overflow_string_threshold - (double)average_length) / (double)overflow_string_threshold) * 25.0;
```

Comment:
Maybe an idea for a benchmark would be to run strings of varying sizes (e.g. 500 bytes, 1000 bytes, etc) and then check at which point zstd starts to perform better, then we could perhaps tune this heuristic along those lines

Reply:
![image](https://github.com/user-attachments/assets/23fa0200-76c2-4d66-925d-5958ba7a2321)
scan speed difference plotted

Reply:
write speed difference plotted
![image](https://github.com/user-attachments/assets/dd58069b-ec62-43c7-9b3d-b290407cb749)

Reply:
I decided to put the penalty to 2x for average_string_length <= 100, 1x for 4096 (the overflow string threshold) and up.
Returning invalid for others, it's much slower than uncompressed and the ratio would likely be between 2x and 4x, making it not worth it